\hypertarget{trim_8c}{\section{src/trim.c File Reference}
\label{trim_8c}\index{src/trim.\+c@{src/trim.\+c}}
}


trims/filter sequences after Quality, N's contaminations.  


{\ttfamily \#include $<$string.\+h$>$}\\*
{\ttfamily \#include $<$stdlib.\+h$>$}\\*
{\ttfamily \#include \char`\"{}trim.\+h\char`\"{}}\\*
{\ttfamily \#include \char`\"{}str\+\_\+manip.\+h\char`\"{}}\\*
{\ttfamily \#include \char`\"{}defines.\+h\char`\"{}}\\*
{\ttfamily \#include \char`\"{}config.\+h\char`\"{}}\\*
Include dependency graph for trim.\+c\+:\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=350pt]{trim_8c__incl}
\end{center}
\end{figure}
\subsection*{Macros}
\begin{DoxyCompactItemize}
\item 
\#define \hyperlink{trim_8c_aaeb5f4dc06b2ca1b46fa5212b6b8a203}{T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G}~20
\end{DoxyCompactItemize}
\subsection*{Functions}
\begin{DoxyCompactItemize}
\item 
static int \hyperlink{trim_8c_aa1ba5f740cba7b39672dc588b9a5deb7}{no\+\_\+\+N} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq)
\begin{DoxyCompactList}\small\item\em checks if a sequence contains any non standard base callings (N's) \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_a7006063272037ceb080150136d24e898}{Nfree\+\_\+\+Lmer} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int min\+L)
\begin{DoxyCompactList}\small\item\em Finds the largest Nfree sub-\/seq and keeps it if larger than min\+L. \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_ae2aeb2e7303c53e2a10009ee4e20e40e}{Ntrim\+\_\+ends} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int min\+L)
\begin{DoxyCompactList}\small\item\em trims a read if N's are at the ends and the remaining sub-\/seq $>$= min\+L \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_ae11fa52547dfdb0cf2af554eafbba1fc}{no\+\_\+low\+Q} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int min\+Q)
\begin{DoxyCompactList}\small\item\em checks if a sequence contains low\+Q nucleotides \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_a03d11622e683d103cec196c42dd654dd}{Qtrim\+\_\+ends} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int min\+Q, int min\+L)
\begin{DoxyCompactList}\small\item\em trims a read if low\+Qs are at the ends and remaining sub-\/seq $>$= min\+L \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_af3167128b208288e71c0c50f15e66e89}{Qtrim\+\_\+frac} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int min\+Q, int nlow\+Q)
\begin{DoxyCompactList}\small\item\em accepts the sequence as is if there are less than nlow\+Q \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_aeef0fcb68dba0ae69ef81bd330a037ad}{Qtrim\+\_\+endsfrac} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int min\+Q, int min\+L, int nlow\+Q)
\begin{DoxyCompactList}\small\item\em trims the ends for low\+Q. The rest is kept if it contains $<$ nlow\+Q low\+Q \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_a1983a90e9b9ca5c39c943c4ea523b6fa}{Qtrim\+\_\+global} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, int left, int right, char type)
\begin{DoxyCompactList}\small\item\em trims left from the left and right from the right \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_a03005207176057abeac4d0ab60afbf93}{align\+\_\+uint32} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, \hyperlink{adapters_8h_ad1465e8f3bc3d3b1fc010a1e72436c56}{Ad\+\_\+seq} $\ast$ptr\+\_\+adap, \hyperlink{defines_8h_abb452686968e48b67397da5f97445f5b}{bool} all)
\begin{DoxyCompactList}\small\item\em alignment search between a fq read, and an adapter sequence, with a seed of 8 nucleotides. \end{DoxyCompactList}\item 
static int \hyperlink{trim_8c_acb771cbd1c3867505712a6f302ec0d2d}{align\+\_\+uint64} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, \hyperlink{adapters_8h_ad1465e8f3bc3d3b1fc010a1e72436c56}{Ad\+\_\+seq} $\ast$ptr\+\_\+adap)
\begin{DoxyCompactList}\small\item\em Alignment search between a fq read, and an adapter sequence, w with a seed of 8 nucleotides. \end{DoxyCompactList}\item 
int \hyperlink{trim_8c_aed995029d7181d69875b2dee0d7b6ce9}{trim\+\_\+adapter} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, \hyperlink{adapters_8h_ad1465e8f3bc3d3b1fc010a1e72436c56}{Ad\+\_\+seq} $\ast$adap\+\_\+list)
\begin{DoxyCompactList}\small\item\em trims sequence based on presence of N nucleotides \end{DoxyCompactList}\item 
int \hyperlink{trim_8c_af23d3685193b624b72a6969a49903a97}{trim\+\_\+sequence\+N} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq)
\begin{DoxyCompactList}\small\item\em trims sequence based on presence of N nucleotides \end{DoxyCompactList}\item 
int \hyperlink{trim_8c_ae5ab2bb79068d8a7e0d72686c1f50320}{trim\+\_\+sequence\+Q} (\hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq)
\begin{DoxyCompactList}\small\item\em trims sequence based on low\+Q base callings \end{DoxyCompactList}\item 
\hyperlink{defines_8h_abb452686968e48b67397da5f97445f5b}{bool} \hyperlink{trim_8c_a7e2b4015298659b7e21a80d4bb031e2f}{is\+\_\+read\+\_\+in\+Tree} (\hyperlink{tree_8h_a50a06950fa1e82738ad9a6bd85914900}{Tree} $\ast$tree\+\_\+ptr, \hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq)
\begin{DoxyCompactList}\small\item\em check if Lread is contained in tree. It computes the score for the read and its reverse complement; if one ot them exceeds the user selected threshold, it returns true. Otherwise, it returns false. \end{DoxyCompactList}\item 
\hyperlink{defines_8h_abb452686968e48b67397da5f97445f5b}{bool} \hyperlink{trim_8c_a6206f2c21429fd2ff36010b71989696d}{is\+\_\+read\+\_\+in\+Bloom} (\hyperlink{bloom_8h_a6cb32cf059e8f4efd1ef80938d982836}{Bfilter} $\ast$ptr\+\_\+bf, \hyperlink{fq__read_8h_a9af37aa81397c9531c66863d4e97f034}{Fq\+\_\+read} $\ast$seq, \hyperlink{bloom_8h_a4bf3f34a321ce545e533d29ce363f569}{Bfkmer} $\ast$ptr\+\_\+bfkmer)
\begin{DoxyCompactList}\small\item\em checks if a read is in Bloom filter. It computes the score for the read and returns true if it exceeds the user selected threshold. Returns false othersise. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Variables}
\begin{DoxyCompactItemize}
\item 
uint8\+\_\+t \hyperlink{trim_8c_a4c94b467d7dec9ee2a642ee537d5e967}{fw\+\_\+1\+B} \mbox{[}256\mbox{]}
\item 
uint8\+\_\+t \hyperlink{trim_8c_a2abce22af1da49bd08cf2baa1d74b939}{Nencode}
\item 
\hyperlink{init__trimFilter_8h_aff83806f1ff2b5f3609dc7d4c04b7c19}{Iparam\+\_\+trim\+Filter} \hyperlink{trim_8c_aed14b08f25fd6aa9a6cffb8c0074c953}{par\+\_\+\+T\+F}
\end{DoxyCompactItemize}


\subsection{Detailed Description}
trims/filter sequences after Quality, N's contaminations. 

\begin{DoxyAuthor}{Author}
Paula Perez \href{mailto:paulaperezrubio@gmail.com}{\tt paulaperezrubio@gmail.\+com} 
\end{DoxyAuthor}
\begin{DoxyDate}{Date}
24.\+08.\+2017 
\end{DoxyDate}


\subsection{Macro Definition Documentation}
\hypertarget{trim_8c_aaeb5f4dc06b2ca1b46fa5212b6b8a203}{\index{trim.\+c@{trim.\+c}!T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G@{T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G}}
\index{T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G@{T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G}!trim.\+c@{trim.\+c}}
\subsubsection[{T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G}]{\setlength{\rightskip}{0pt plus 5cm}\#define T\+R\+I\+M\+\_\+\+S\+T\+R\+I\+N\+G~20}}\label{trim_8c_aaeb5f4dc06b2ca1b46fa5212b6b8a203}
maximal length of trimming info string. 

\subsection{Function Documentation}
\hypertarget{trim_8c_a03005207176057abeac4d0ab60afbf93}{\index{trim.\+c@{trim.\+c}!align\+\_\+uint32@{align\+\_\+uint32}}
\index{align\+\_\+uint32@{align\+\_\+uint32}!trim.\+c@{trim.\+c}}
\subsubsection[{align\+\_\+uint32}]{\setlength{\rightskip}{0pt plus 5cm}static int align\+\_\+uint32 (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{{\bf Ad\+\_\+seq} $\ast$}]{ptr\+\_\+adap, }
\item[{{\bf bool}}]{all}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_a03005207176057abeac4d0ab60afbf93}


alignment search between a fq read, and an adapter sequence, with a seed of 8 nucleotides. 

This function checks whether there is adapter contamination in a given read. It works stand alone if the adapter is shorter than 16 nucleotides, and is called from align\+\_\+uint64 when no 16-\/nucleotides long seeds are found. The criteria are the same as in align\+\_\+uint64, the seed length being 8-\/nucleotides long instead of 16. See the {\bfseries align\+\_\+uint64} documentation for more details.


\begin{DoxyParams}{Parameters}
{\em seq} & pointer to {\bfseries Fq\+\_\+read} \\
\hline
{\em adap\+\_\+list} & array of {\bfseries Ad\+\_\+seq} \\
\hline
{\em all} & true if the whole read has to be sweeped, false if only the ends. When this function is called from align\+\_\+uint64, only the ends need to be considered. \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
-\/1 error, 0 discarded, 1 accepted as is, 2 accepted and trimmed 
\end{DoxyReturn}
\begin{DoxyNote}{Note}
Global input parameters from par\+\_\+\+T\+F are also used 
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{init__trimFilter_8h_a3442d07a10a4ab75064687f3c244e370}{Adapter} 

\hyperlink{init__trimFilter_8h_aff83806f1ff2b5f3609dc7d4c04b7c19}{Iparam\+\_\+trim\+Filter} 

\hyperlink{trim_8c_acb771cbd1c3867505712a6f302ec0d2d}{align\+\_\+uint64} 

\hyperlink{adapters_8h_a474bb2a439e8396fe33f6998af243eee}{pack\+\_\+adapter} 

\hyperlink{adapters_8h_adc47abb63194fec1a3ec9d83991a1340}{obtain\+\_\+score} 
\end{DoxySeeAlso}
\hypertarget{trim_8c_acb771cbd1c3867505712a6f302ec0d2d}{\index{trim.\+c@{trim.\+c}!align\+\_\+uint64@{align\+\_\+uint64}}
\index{align\+\_\+uint64@{align\+\_\+uint64}!trim.\+c@{trim.\+c}}
\subsubsection[{align\+\_\+uint64}]{\setlength{\rightskip}{0pt plus 5cm}static int align\+\_\+uint64 (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{{\bf Ad\+\_\+seq} $\ast$}]{ptr\+\_\+adap}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_acb771cbd1c3867505712a6f302ec0d2d}


Alignment search between a fq read, and an adapter sequence, w with a seed of 8 nucleotides. 


\begin{DoxyParams}{Parameters}
{\em seq} & pointer to {\bfseries Fq\+\_\+read} \\
\hline
{\em adap\+\_\+list} & array of {\bfseries Ad\+\_\+seq} \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
-\/1 error, 0 discarded, 1 accepted as is, 2 accepted and trimmed 
\end{DoxyReturn}
\begin{DoxyNote}{Note}
Global input parameters from par\+\_\+\+T\+F are used 
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{init__trimFilter_8h_a3442d07a10a4ab75064687f3c244e370}{Adapter} 

\hyperlink{init__trimFilter_8h_aff83806f1ff2b5f3609dc7d4c04b7c19}{Iparam\+\_\+trim\+Filter} 

\hyperlink{trim_8c_a03005207176057abeac4d0ab60afbf93}{align\+\_\+uint32} 

\hyperlink{adapters_8h_a474bb2a439e8396fe33f6998af243eee}{pack\+\_\+adapter} 

\hyperlink{adapters_8h_adc47abb63194fec1a3ec9d83991a1340}{obtain\+\_\+score}
\end{DoxySeeAlso}
This function checks whether there is adapter contamination in a given read. We start by looking for 16-\/nucleotides long seeds, where a user defined number of mismatches is allowed. If found, a score is computed. If the score is larger than the user defined threshold and the number of matched nucleotides exceeds M\+I\+N\+\_\+\+N\+M\+A\+T\+C\+H\+E\+S (12), then the read is trimmed if the remaining part is longer than min\+L (user defined) and discarded otherwise. If no 16-\/nucleotides long seeds are found, we proceed with 8-\/nucleotides long seeds (see align\+\_\+uint64) and apply the same criteria to trim/discard a read. A list of possible situations follows, to illustrate how it works (min\+L=25, mismatches=2)\+:


\begin{DoxyCode}
1 ADAPTER: CAAGCAGAAGACGGCATACGAG
2 REV\_COM: AGATCGGAAGAGCTCGTATGCC
3 
4 CASE1A:  CACAGTCGATCAGCGAGCAGGCATTCATGCTGAGATCGGAAGAGATCGTATG
5                                          ||||||||||||X|||----
6                                          AGATCGGAAGAGCTCGTATG
7          - Seed: 16 Nucleotides
8          - Return: 2, TRIMA:0:31
9 CASE1B:  CACATCATCGCTAGCTATCGATCGATCGATGCTATGCAAGATCGGAAGAGCT
10                                                ||||||||------
11                                                AGATCGGAAGAGCT
12          - Seed: 8 Nucleotides
13          - Return: 2, TRIMA:0:37
14 CASE1C:  CACATCATCGCTAGCTATCGATCGATCGATGCTATGCACGAAGATCGGAAGA
15                                                   ||||||||---
16                                                   AGATCGGAAGA
17          - Seed: 8 Nucleotides
18          - Return: 1, reason: Match length < 12
19 CASE2A:  CATACATCACGAGCTAGCTAGAGATCGGAAGAGCTCGTATGCCCAGCATCGA
20                                ||||||||||||||||------
21                                AGATCGGAAGAGCTCGTATGCC
22          - Seed: 16 Nucleotides
23          - Return: 0, reason: remaining read too short.
24 CASE2B:  CCACAGTACAATACATCACGAGCTAGCTAGAGATCGGAAGAGCTCGTATGCA
25                                      ||||||||||||||||||||||
26                                      AGATCGGAAGAGCTCGTATGCC
27          - Seed: 16 Nucleotides
28          - Return: 2, TRIMA:0:28
29 CASE3A:  TATGCCGTCTTCTGCTTGCAGTGCATGCTGATGCATGCTGCATGCTAGCTGC
30          ||||||||||||||||--
31          TATGCCGTCTTCTGCTTG
32          - Seed: 16 Nucleotides
33          - Return: 0, reason: remaining read too short
34 CASE3B:  CGTCTTCTGCTTGCCGATCGATGCTAGCTACGATCGTCGAGCTAGCTACGTG
35          ||||||||-----
36          CGTCTTCTGCTTG
37          - Seed: 8 Nucleotides
38          - Return: 0, reason: remaining read too short
39 CASE3C:  TCTTCTGCTTGCCGATCGATGCTAGCTACGATCGTCGAGCTAGCTACGTGCG
40          ||||||||---
41          TCTTCTGCTTG
42          - Seed: 8 Nucleotides
43          - Return: 1, reason: Match length < 12
\end{DoxyCode}
 \hypertarget{trim_8c_a6206f2c21429fd2ff36010b71989696d}{\index{trim.\+c@{trim.\+c}!is\+\_\+read\+\_\+in\+Bloom@{is\+\_\+read\+\_\+in\+Bloom}}
\index{is\+\_\+read\+\_\+in\+Bloom@{is\+\_\+read\+\_\+in\+Bloom}!trim.\+c@{trim.\+c}}
\subsubsection[{is\+\_\+read\+\_\+in\+Bloom}]{\setlength{\rightskip}{0pt plus 5cm}{\bf bool} is\+\_\+read\+\_\+in\+Bloom (
\begin{DoxyParamCaption}
\item[{{\bf Bfilter} $\ast$}]{ptr\+\_\+bf, }
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{{\bf Bfkmer} $\ast$}]{ptr\+\_\+bfkmer}
\end{DoxyParamCaption}
)}}\label{trim_8c_a6206f2c21429fd2ff36010b71989696d}


checks if a read is in Bloom filter. It computes the score for the read and returns true if it exceeds the user selected threshold. Returns false othersise. 


\begin{DoxyParams}{Parameters}
{\em ptr\+\_\+bf} & pointer to Bfilter \\
\hline
{\em seq} & fastq read \\
\hline
{\em ptr\+\_\+bfkmer} & pointer to Procs\+\_\+kmer structure (will store global) \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
true if read was found, false otherwise 
\end{DoxyReturn}
\hypertarget{trim_8c_a7e2b4015298659b7e21a80d4bb031e2f}{\index{trim.\+c@{trim.\+c}!is\+\_\+read\+\_\+in\+Tree@{is\+\_\+read\+\_\+in\+Tree}}
\index{is\+\_\+read\+\_\+in\+Tree@{is\+\_\+read\+\_\+in\+Tree}!trim.\+c@{trim.\+c}}
\subsubsection[{is\+\_\+read\+\_\+in\+Tree}]{\setlength{\rightskip}{0pt plus 5cm}{\bf bool} is\+\_\+read\+\_\+in\+Tree (
\begin{DoxyParamCaption}
\item[{{\bf Tree} $\ast$}]{tree\+\_\+ptr, }
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq}
\end{DoxyParamCaption}
)}}\label{trim_8c_a7e2b4015298659b7e21a80d4bb031e2f}


check if Lread is contained in tree. It computes the score for the read and its reverse complement; if one ot them exceeds the user selected threshold, it returns true. Otherwise, it returns false. 


\begin{DoxyParams}{Parameters}
{\em tree\+\_\+ptr} & pointer to Tree structure \\
\hline
{\em seq} & fastq read \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
true if read was found, false otherwise 
\end{DoxyReturn}
\hypertarget{trim_8c_a7006063272037ceb080150136d24e898}{\index{trim.\+c@{trim.\+c}!Nfree\+\_\+\+Lmer@{Nfree\+\_\+\+Lmer}}
\index{Nfree\+\_\+\+Lmer@{Nfree\+\_\+\+Lmer}!trim.\+c@{trim.\+c}}
\subsubsection[{Nfree\+\_\+\+Lmer}]{\setlength{\rightskip}{0pt plus 5cm}static int Nfree\+\_\+\+Lmer (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{min\+L}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_a7006063272037ceb080150136d24e898}


Finds the largest Nfree sub-\/seq and keeps it if larger than min\+L. 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em min\+L} & minimum accepted trimmed length \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
0 if not used, 1 if accepted as is, 2 if accepted and trimmed 
\end{DoxyReturn}
\hypertarget{trim_8c_ae11fa52547dfdb0cf2af554eafbba1fc}{\index{trim.\+c@{trim.\+c}!no\+\_\+low\+Q@{no\+\_\+low\+Q}}
\index{no\+\_\+low\+Q@{no\+\_\+low\+Q}!trim.\+c@{trim.\+c}}
\subsubsection[{no\+\_\+low\+Q}]{\setlength{\rightskip}{0pt plus 5cm}static int no\+\_\+low\+Q (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{min\+Q}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_ae11fa52547dfdb0cf2af554eafbba1fc}


checks if a sequence contains low\+Q nucleotides 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em min\+Q} & minimum accepted quality value \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
0 if seq contains low\+Q nucleotides, 1 otherwise 
\end{DoxyReturn}
\hypertarget{trim_8c_aa1ba5f740cba7b39672dc588b9a5deb7}{\index{trim.\+c@{trim.\+c}!no\+\_\+\+N@{no\+\_\+\+N}}
\index{no\+\_\+\+N@{no\+\_\+\+N}!trim.\+c@{trim.\+c}}
\subsubsection[{no\+\_\+\+N}]{\setlength{\rightskip}{0pt plus 5cm}static int no\+\_\+\+N (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_aa1ba5f740cba7b39672dc588b9a5deb7}


checks if a sequence contains any non standard base callings (N's) 

\begin{DoxyReturn}{Returns}
0 if no N's found, 1 if N's found
\end{DoxyReturn}
This function checks if any of the base callings in a given fastq read is different from A, C, G, T. Basically, any char different from the former ones is classified as N. \hypertarget{trim_8c_ae2aeb2e7303c53e2a10009ee4e20e40e}{\index{trim.\+c@{trim.\+c}!Ntrim\+\_\+ends@{Ntrim\+\_\+ends}}
\index{Ntrim\+\_\+ends@{Ntrim\+\_\+ends}!trim.\+c@{trim.\+c}}
\subsubsection[{Ntrim\+\_\+ends}]{\setlength{\rightskip}{0pt plus 5cm}static int Ntrim\+\_\+ends (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{min\+L}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_ae2aeb2e7303c53e2a10009ee4e20e40e}


trims a read if N's are at the ends and the remaining sub-\/seq $>$= min\+L 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em min\+L} & minimum accepted trimmed length \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
0 if not used, 1 no N's found, 2 if accepted and trimmed 
\end{DoxyReturn}
\hypertarget{trim_8c_a03d11622e683d103cec196c42dd654dd}{\index{trim.\+c@{trim.\+c}!Qtrim\+\_\+ends@{Qtrim\+\_\+ends}}
\index{Qtrim\+\_\+ends@{Qtrim\+\_\+ends}!trim.\+c@{trim.\+c}}
\subsubsection[{Qtrim\+\_\+ends}]{\setlength{\rightskip}{0pt plus 5cm}static int Qtrim\+\_\+ends (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{min\+Q, }
\item[{int}]{min\+L}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_a03d11622e683d103cec196c42dd654dd}


trims a read if low\+Qs are at the ends and remaining sub-\/seq $>$= min\+L 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em min\+Q} & minimum accepted quality value \\
\hline
{\em min\+L} & minimum accepted trimmed length \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
0 if not used, 1 if accepted as is, 2 if accepted and trimmed 
\end{DoxyReturn}
\hypertarget{trim_8c_aeef0fcb68dba0ae69ef81bd330a037ad}{\index{trim.\+c@{trim.\+c}!Qtrim\+\_\+endsfrac@{Qtrim\+\_\+endsfrac}}
\index{Qtrim\+\_\+endsfrac@{Qtrim\+\_\+endsfrac}!trim.\+c@{trim.\+c}}
\subsubsection[{Qtrim\+\_\+endsfrac}]{\setlength{\rightskip}{0pt plus 5cm}static int Qtrim\+\_\+endsfrac (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{min\+Q, }
\item[{int}]{min\+L, }
\item[{int}]{nlow\+Q}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_aeef0fcb68dba0ae69ef81bd330a037ad}


trims the ends for low\+Q. The rest is kept if it contains $<$ nlow\+Q low\+Q 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em min\+Q} & minimum accepted quality value \\
\hline
{\em min\+L} & minimum accepted trimmed length \\
\hline
{\em nlow\+Q} & threshold on low\+Q nucleotides ($>$= N\+O\+T allowed) \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
0 if not used, 1 if accepted as is
\end{DoxyReturn}
Trims the ends if there are low\+Q bases. From the remaining part, it counts how many low\+Q bases there are and keeps it if there are less than nlow\+Q. \hypertarget{trim_8c_af3167128b208288e71c0c50f15e66e89}{\index{trim.\+c@{trim.\+c}!Qtrim\+\_\+frac@{Qtrim\+\_\+frac}}
\index{Qtrim\+\_\+frac@{Qtrim\+\_\+frac}!trim.\+c@{trim.\+c}}
\subsubsection[{Qtrim\+\_\+frac}]{\setlength{\rightskip}{0pt plus 5cm}static int Qtrim\+\_\+frac (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{min\+Q, }
\item[{int}]{nlow\+Q}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_af3167128b208288e71c0c50f15e66e89}


accepts the sequence as is if there are less than nlow\+Q 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em min\+Q} & minimum accepted quality value \\
\hline
{\em nlow\+Q} & threshold on low\+Q nucleotides ($>$= N\+O\+T allowed) \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
0 if not used, 1 if accepted as is 
\end{DoxyReturn}
\hypertarget{trim_8c_a1983a90e9b9ca5c39c943c4ea523b6fa}{\index{trim.\+c@{trim.\+c}!Qtrim\+\_\+global@{Qtrim\+\_\+global}}
\index{Qtrim\+\_\+global@{Qtrim\+\_\+global}!trim.\+c@{trim.\+c}}
\subsubsection[{Qtrim\+\_\+global}]{\setlength{\rightskip}{0pt plus 5cm}static int Qtrim\+\_\+global (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{int}]{left, }
\item[{int}]{right, }
\item[{char}]{type}
\end{DoxyParamCaption}
)\hspace{0.3cm}{\ttfamily [static]}}}\label{trim_8c_a1983a90e9b9ca5c39c943c4ea523b6fa}


trims left from the left and right from the right 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
{\em left} & number of nucleotides to be trimmed from the left \\
\hline
{\em right} & number of nucleotides to be trimmed from the right \\
\hline
{\em type} & char indicating the type of trimming (Q,A). \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
2, since they are all accepted and trim 
\end{DoxyReturn}
\hypertarget{trim_8c_aed995029d7181d69875b2dee0d7b6ce9}{\index{trim.\+c@{trim.\+c}!trim\+\_\+adapter@{trim\+\_\+adapter}}
\index{trim\+\_\+adapter@{trim\+\_\+adapter}!trim.\+c@{trim.\+c}}
\subsubsection[{trim\+\_\+adapter}]{\setlength{\rightskip}{0pt plus 5cm}int trim\+\_\+adapter (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq, }
\item[{{\bf Ad\+\_\+seq} $\ast$}]{adap\+\_\+list}
\end{DoxyParamCaption}
)}}\label{trim_8c_aed995029d7181d69875b2dee0d7b6ce9}


trims sequence based on presence of N nucleotides 

if (adapter length $<$ 16) -\/$>$ search for seeds 8 nucleotides long else -\/$>$ search for seeds 16 nucleotides long if (seed found) -\/$>$ calculate score if score $>$ threshold -\/$>$ aligner found, trim / discard and exit. else -\/$>$ search for seeds 8 nucleotides long 
\begin{DoxyParams}{Parameters}
{\em seq} & pointer to {\bfseries Fq\+\_\+read} \\
\hline
{\em adap\+\_\+list} & array of {\bfseries Ad\+\_\+seq} \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
-\/1 error, 0 discarded, 1 accepted as is, 2 accepted and trimmed 
\end{DoxyReturn}
\begin{DoxyNote}{Note}
Global input parameters from par\+\_\+\+T\+F are also used 
\end{DoxyNote}
\hypertarget{trim_8c_af23d3685193b624b72a6969a49903a97}{\index{trim.\+c@{trim.\+c}!trim\+\_\+sequence\+N@{trim\+\_\+sequence\+N}}
\index{trim\+\_\+sequence\+N@{trim\+\_\+sequence\+N}!trim.\+c@{trim.\+c}}
\subsubsection[{trim\+\_\+sequence\+N}]{\setlength{\rightskip}{0pt plus 5cm}int trim\+\_\+sequence\+N (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq}
\end{DoxyParamCaption}
)}}\label{trim_8c_af23d3685193b624b72a6969a49903a97}


trims sequence based on presence of N nucleotides 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
-\/1 error, 0 discarded, 1 accepted as is, 2 accepted and trimmed
\end{DoxyReturn}
This function calls a different function depending on the method passed as input par\+\_\+\+T\+F.\+trim\+N\+:
\begin{DoxyItemize}
\item \hyperlink{defines_8h_a996bde01ecac342918f0a2c4e7ce7bd5}{N\+O(0)}\+: accepts it as is, (1),
\item \hyperlink{defines_8h_a1edd1ea8bddaf4d9c5eb3eae1ee1726a}{A\+L\+L(1)}\+: accepts it as is if N\+O N's found (1), rejects it otherwise (0),
\item \hyperlink{defines_8h_a052e72209cfac2ff9aa78294f0bebea8}{E\+N\+D\+S(2)}\+: trims the ends and accepts it if it is longer than min\+L (2 if trimming, 1 if no trimming), rejects it otherwise (0),
\item \hyperlink{defines_8h_a53529d9638a1d70f6e5989dedf4c2672}{S\+T\+R\+I\+P(3)}\+: finds the longest N-\/free subsequence and trims it if it is at least min\+L nucleotides long (2 if trimming, 1 if no N's are found), rejects it otherwise (0). 
\end{DoxyItemize}\hypertarget{trim_8c_ae5ab2bb79068d8a7e0d72686c1f50320}{\index{trim.\+c@{trim.\+c}!trim\+\_\+sequence\+Q@{trim\+\_\+sequence\+Q}}
\index{trim\+\_\+sequence\+Q@{trim\+\_\+sequence\+Q}!trim.\+c@{trim.\+c}}
\subsubsection[{trim\+\_\+sequence\+Q}]{\setlength{\rightskip}{0pt plus 5cm}int trim\+\_\+sequence\+Q (
\begin{DoxyParamCaption}
\item[{{\bf Fq\+\_\+read} $\ast$}]{seq}
\end{DoxyParamCaption}
)}}\label{trim_8c_ae5ab2bb79068d8a7e0d72686c1f50320}


trims sequence based on low\+Q base callings 


\begin{DoxyParams}{Parameters}
{\em seq} & fastq read \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
-\/1 error, 0 discarded, 1 accepted as is, 2 accepted and trimmed
\end{DoxyReturn}
This function calls a different function depending on the method passed as input par\+\_\+\+T\+F.\+trim\+Q\+:
\begin{DoxyItemize}
\item \hyperlink{defines_8h_a996bde01ecac342918f0a2c4e7ce7bd5}{N\+O(0)}\+: accepts is as is , (1),
\item \hyperlink{defines_8h_a653af6bd29f56a2699de26a928820da7}{F\+R\+A\+C(1)}\+: accepts it if less than par\+\_\+\+T\+F.\+nlow\+Q are found (1), rejects it otherwise (0),
\item \hyperlink{defines_8h_a052e72209cfac2ff9aa78294f0bebea8}{E\+N\+D\+S(2)}\+: trims the ends and accepts it if it is longer than min\+L (2 if triming, 1 if no trimming), rejects it otherwise (0),
\item \hyperlink{defines_8h_abf0b71573c7ffc4f6746c24c9abc202a}{E\+N\+D\+S\+F\+R\+A\+C(3)}\+: trims the ends and accepts if the remaining sequence is at least min\+L bases long and if it contains less than nlow\+Q low\+Q nucleotides (2 if trimming, 1 if no trimming). Otherwise, it is rejected, (0).
\item \hyperlink{defines_8h_a3de33738fd3c7e77bffbcfaefc3e7645}{G\+L\+O\+B\+A\+L(4)}\+: it trims globally globleft nucleotides from the left and globright from the right, (returns 2). 
\end{DoxyItemize}

\subsection{Variable Documentation}
\hypertarget{trim_8c_a4c94b467d7dec9ee2a642ee537d5e967}{\index{trim.\+c@{trim.\+c}!fw\+\_\+1\+B@{fw\+\_\+1\+B}}
\index{fw\+\_\+1\+B@{fw\+\_\+1\+B}!trim.\+c@{trim.\+c}}
\subsubsection[{fw\+\_\+1\+B}]{\setlength{\rightskip}{0pt plus 5cm}uint8\+\_\+t fw\+\_\+1\+B\mbox{[}256\mbox{]}}}\label{trim_8c_a4c94b467d7dec9ee2a642ee537d5e967}
global variable. Lookup table. D\+On't need it

global variable. Lookup table. \hypertarget{trim_8c_a2abce22af1da49bd08cf2baa1d74b939}{\index{trim.\+c@{trim.\+c}!Nencode@{Nencode}}
\index{Nencode@{Nencode}!trim.\+c@{trim.\+c}}
\subsubsection[{Nencode}]{\setlength{\rightskip}{0pt plus 5cm}uint8\+\_\+t Nencode}}\label{trim_8c_a2abce22af1da49bd08cf2baa1d74b939}
global variable. Encoding for N's(\textbackslash{}004, or \textbackslash{}005) \hypertarget{trim_8c_aed14b08f25fd6aa9a6cffb8c0074c953}{\index{trim.\+c@{trim.\+c}!par\+\_\+\+T\+F@{par\+\_\+\+T\+F}}
\index{par\+\_\+\+T\+F@{par\+\_\+\+T\+F}!trim.\+c@{trim.\+c}}
\subsubsection[{par\+\_\+\+T\+F}]{\setlength{\rightskip}{0pt plus 5cm}{\bf Iparam\+\_\+trim\+Filter} par\+\_\+\+T\+F}}\label{trim_8c_aed14b08f25fd6aa9a6cffb8c0074c953}
global variable\+: Input parameters trim\+Filter.

global variable\+: Input parameters of make\+Tree. 