<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>My Project: trimFilter user manual</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">My Project
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">trimFilter user manual </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>This program reads a <code>fastq</code> file as an input, and filters it according to the following criteria:</p><ul>
<li>Discard/trims reads containing adapter remnants.</li>
<li>Discards reads matching contaminations (sequences collected in a <code>fasta</code> file or in an <code>idx</code> file created by <code>makeTree</code>, <code>makeBloom</code></li>
<li>Discards/trims low quality reads.</li>
<li>Discards/trims reads containing N base callings.</li>
</ul>
<h2>Running the program</h2>
<p>Usage <code>C</code> executable (in folder <code>bin</code>):</p>
<div class="fragment"><div class="line">Usage: trimFilter --ifq &lt;INPUT_FILE.fq&gt; --length &lt;READ_LENGTH&gt;</div><div class="line">q                  --output [O_PREFIX] --gzip [y|n]</div><div class="line">                  --adapter [&lt;ADAPTERS.fa&gt;:&lt;mismatches&gt;:&lt;score&gt;]</div><div class="line">                  --method [TREE|BLOOM]</div><div class="line">                  (--idx [&lt;INDEX_FILE&gt;:&lt;score&gt;:&lt;lmer_len&gt;] |</div><div class="line">                   --ifa [&lt;INPUT.fa&gt;:&lt;score&gt;:[lmer_len]])</div><div class="line">                  --trimQ [NO|ALL|ENDS|FRAC|ENDSFRAC|GLOBAL]</div><div class="line">                  --minL [MINL]  --minQ [MINQ]</div><div class="line">                  (--percent [percent] | --global [n1:n2])</div><div class="line">                  --trimN [NO|ALL|ENDS|STRIP]</div><div class="line">Reads in a fq file (gz, bz2, z formats also accepted) and removes:</div><div class="line">  * low quality reads,</div><div class="line">  * reads containing N base callings,</div><div class="line">  * reads representing contaminations, belonging to sequences in INPUT.fa</div><div class="line">Outputs 4 [O_PREFIX]_fq.gz files containing: &quot;good&quot; reads, discarded</div><div class="line">low Q reads discarded reads containing N&#39;s, discarded contaminations.</div><div class="line">Options:</div><div class="line"> -v, --version prints package version.</div><div class="line"> -h, --help    prints help dialog.</div><div class="line"> -f, --ifq     fastq input file [*fq|*fq.gz|*fq.bz2], mandatory option.</div><div class="line"> -l, --length  read length: length of the reads, mandatory option.</div><div class="line"> -o, --output  output prefix (with path), optional (default ./out).</div><div class="line"> -z, --gzip    gzips output files: yes or no (default yes).</div><div class="line"> -A, --adapter adapter input three fields separated by colons:</div><div class="line">               &lt;ADAPTERS.fa&gt;: fasta file containing adapters,</div><div class="line">               &lt;mismatches&gt;: maximum mismatch count allowed,</div><div class="line">               &lt;score&gt;: score threshold  for the aligner.</div><div class="line"> -x, --idx     index input file. To be included with any method. 3 fields</div><div class="line">               3 fields separated by colons:</div><div class="line">               &lt;INDEX_FILE&gt;: output of makeTree, makeBloom,</div><div class="line">               &lt;score&gt;: score threshold to accept a match [0,1],</div><div class="line">               [lmer_len]: correspond to the length of the lmers to be</div><div class="line">                        looked for in the reads [1,READ_LENGTH].</div><div class="line"> -a, --ifa     fasta input file. To be included only with method TREE</div><div class="line">               (it excludes the option --idx). Otherwise, an</div><div class="line">               index file has to be precomputed and given as parameter</div><div class="line">               (see option --idx). 3 fields separated by colons:</div><div class="line">               &lt;INPUT.fa&gt;: fasta input file [*fa|*fa.gz|*fa.bz2],</div><div class="line">               &lt;score&gt;: score threshold to accept a match [0,1],</div><div class="line">               &lt;lmer_len&gt;: depth of the tree: [1,READ_LENGTH]. It will</div><div class="line">                        correspond to the length of the lmers to be</div><div class="line">                        looked for in the reads.</div><div class="line"> -C, --method  method used to look for contaminations:</div><div class="line">               TREE:  uses a 4-ary tree. Index file optional,</div><div class="line">               BLOOM: uses a bloom filter. Index file mandatory.</div><div class="line"> -Q, --trimQ   NO:       does nothing to low quality reads (default),</div><div class="line">               ALL:      removes all reads containing at least one low</div><div class="line">                         quality nucleotide.</div><div class="line">               ENDS:     trims the ends of the read if their quality is</div><div class="line">                         below the threshold -q,</div><div class="line">               FRAC:     discards a read if the fraction of bases whose</div><div class="line">                         quality lies below</div><div class="line">                         the threshold -q is over 5 percent or a user</div><div class="line">                         defined percentage in -p.</div><div class="line">               ENDSFRAC: trims the ends and then discards the read if</div><div class="line">                         there are more low quality nucleotides than the</div><div class="line">                         allowed by the option -p.</div><div class="line">               GLOBAL:   removes n1 cycles on the left and n2 on the</div><div class="line">                         right, specified in -g.</div><div class="line">               All reads are discarded if they are shorter than MINL.</div><div class="line"> -m, --minL    minimum length allowed for a read before it is discarded</div><div class="line">               (default 25).</div><div class="line"> -q, --minQ    minimum quality allowed (int), optional (default 27).</div><div class="line"> -p, --percent percentage of low quality bases to be admitted before</div><div class="line">               discarding a read (default 5),</div><div class="line"> -g, --global  required option if --trimQ GLOBAL is passed. Two int,</div><div class="line">               n1:n2, have to be passed specifying the number of cycles</div><div class="line">               to be globally cut from the left and right, respectively.</div><div class="line"> -N, --trimN   NO:     does nothing to reads containing N&#39;s,</div><div class="line">               ALL:    removes all reads containing N&#39;s,</div><div class="line">               ENDS:   trims ends of reads with N&#39;s,</div><div class="line">               STRIPS: looks for the largest substring with no N&#39;s.</div><div class="line">               All reads are discarded if they are shorter than minL.</div></div><!-- fragment --><p>NOTE: the parameters -l or &ndash;length are meant to identify the length of the reads in the input data. Actually, <code>trimFilter</code> also copes with data holding reads with different lengths. The length parameter must hold the length of the longest read in the dataset.</p>
<h2>Output description</h2>
<ul>
<li><code>O_PREFIX_good.fq.gz</code>: contains reads that passed all filters (maybe trimmed).</li>
<li><code>O_PREFIX_adap.fq.gz</code>: contains reads discarded due to the presence of adapters.</li>
<li><code>O_PREFIX_cont.fq.gz</code>: contains contamination reads.</li>
<li><code>O_PREFIX_lowQ.fq.gz</code>: contains reads discarded due to low quality issues.</li>
<li><code>O_PREFIX_NNNN.fq.gz</code>: contains reads discarded due to <em>N</em>'s issues.</li>
<li><code>O_PREFIX_summary.bin</code>: binary file where information about the filtering process is stored. Structure of the file.<ul>
<li>filters, <code>4*sizeof(int) Bytes</code>: array of int with entries <code>i = {ADAP(0), CONT(1), LOWQ(2), NNNN(3)}</code>. A given entry takes the value of the filter it was applied to and 0 otherwise. <code>filters[ADAPT] = {0,1}</code>, <code>filters[CONT] = {NO(0), TREE(1), BLOOM(2)}</code>, <br />
 <code>filters[LOWQ] = {NO(0), ALL(1), ENDS(2), FRAC(3), ENDSFRAC(4), GLOBAL(5)}</code>, <code>filters[trimN] = {NO(0), ALL(1), ENDS(2), STRIPS(2)}</code>.</li>
<li>trimmed, <code>4*sizeof(int) Bytes</code>: array of integers with entries i = {ADAP(0), CONT(1), LOWQ(2), NNNN(3)}, containing how many reads were trimmed due to the corresponding filter.</li>
<li>discarded, <code>4*sizeof(int) Bytes</code>: array of integers with entries i = {ADAP(0), CONT(1), LOWQ(2), NNNN(3)}, containing how many reads were discarded due to the corresponding filter.</li>
<li>good, <code>sizeof(int) Bytes</code>: number of accepted reads (maybe trimmed).</li>
<li>nreads, <code>sizeof(int) Bytes</code>: total number of reads.</li>
</ul>
</li>
</ul>
<h2>Filters</h2>
<h4>Adapters</h4>
<p>Technical sequences within the reads are detected if the option <code>--adapters &lt;ADAPTERS.fa&gt;:&lt;mismatches&gt;:&lt;score&gt;</code> is given. The adapter(s) sequence(s) are read from the fasta file, and the search is done using an 'seed and extend' approach. It starts by looking for 16-nucleotides long seeds, for which a user defined number of mismatches is allowed (<code>mismatches</code>). If found, a score is computed. If the score is larger than the user defined threshold (<code>score</code>) and the number of matched nucleotides exceeds 12, then the read is trimmed if the remaining part is longer than <code>MINL</code> (user defined) and discarded otherwise. If no 16-nucleotides long seeds are found, we proceed with 8-nucleotides long seeds and apply the same criteria to trim/discard a read. A list of possible situations follows, to illustrate how it works (<code>MINL=25</code>, <code>mismatches=2</code>):</p>
<div class="fragment"><div class="line">ADAPTER: GGCATACGAGCTCTTCCGATCT</div><div class="line">REV_COM: AGATCGGAAGAGCTCGTATGCC</div><div class="line"></div><div class="line">CASE1A:  CACAGTCGATCAGCGAGCAGGCATTCATGCTGAGATCGGAAGAGATCGTATG</div><div class="line">                                         ||||||||||||X|||----</div><div class="line">                                         AGATCGGAAGAGCTCGTATG</div><div class="line">         - Seed: 16 Nucleotides</div><div class="line">         - Return: trimmed, TRIMA:0:31</div><div class="line">CASE1B:  CACATCATCGCTAGCTATCGATCGATCGATGCTATGCAAGATCGGAAGAGCT</div><div class="line">                                               ||||||||------</div><div class="line">                                               AGATCGGAAGAGCT</div><div class="line">         - Seed: 8 Nucleotides</div><div class="line">         - Return: trimmed, TRIMA:0:37</div><div class="line">CASE1C:  CACATCATCGCTAGCTATCGATCGATCGATGCTATGCACGAAGATCGGAAGA</div><div class="line">                                                  ||||||||---</div><div class="line">                                                  AGATCGGAAGA</div><div class="line">         - Seed: 8 Nucleotides</div><div class="line">         - Return: nothing done, reason: Match length &lt; 12</div><div class="line">CASE2A:  CATACATCACGAGCTAGCTAGAGATCGGAAGAGCTCGTATGCCCAGCATCGA</div><div class="line">                              ||||||||||||||||------</div><div class="line">                              AGATCGGAAGAGCTCGTATGCC</div><div class="line">         - Seed: 16 Nucleotides</div><div class="line">         - Return: discarded, reason: remaining read too short.</div><div class="line">CASE2B:  CCACAGTACAATACATCACGAGCTAGCTAGAGATCGGAAGAGCTCGTATGCC</div><div class="line">                                       ||||||||||||||||||||||</div><div class="line">                                       AGATCGGAAGAGCTCGTATGCC</div><div class="line">         - Seed: 16 Nucleotides</div><div class="line">         - Return: trimmed, TRIMA:0:28</div><div class="line">CASE3A:  TATGCCGTCTTCTGCTTGCAGTGCATGCTGATGCATGCTGCATGCTAGCTGC</div><div class="line">         ||||||||||||||||--</div><div class="line">         TATGCCGTCTTCTGCTTG</div><div class="line">         - Seed: 16 Nucleotides</div><div class="line">         - Return: discarded, reason: remaining read too short</div><div class="line">CASE3B:  CGTCTTCTGCTTGCCGATCGATGCTAGCTACGATCGTCGAGCTAGCTACGTG</div><div class="line">         ||||||||-----</div><div class="line">         CGTCTTCTGCTTG</div><div class="line">         - Seed: 8 Nucleotides</div><div class="line">         - Return: discarded, reason: remaining read too short</div><div class="line">CASE3C:  TCTTCTGCTTGCCGATCGATGCTAGCTACGATCGTCGAGCTAGCTACGTGCG</div><div class="line">         ||||||||---</div><div class="line">         TCTTCTGCTTG</div><div class="line">         - Seed: 8 Nucleotides</div><div class="line">         - Return: nothing done, reason: Match length &lt; 12</div></div><!-- fragment --><p>The score is calculated as follows:</p><ul>
<li>- matching bases: <code>score += log_10(4)</code></li>
<li>- unmatching bases: <code>score -= Q/10</code>, where Q is the quality score.</li>
</ul>
<h4>Impurities</h4>
<p>Contaminations are removed if a fasta or an index file are given as an input. Three methods have been implemented to check for contaminations:</p>
<ul>
<li><b>TREE</b>: this method is designed to identify impurities from small sequences, such as rRNA, or EColi (not larger than 10MB). When using this method, ont has to pass one of these two options:<ul>
<li><code>--ifa &lt;INPUT.fa&gt;:&lt;score&gt;:&lt;lmer_len&gt;</code>: the file <code>INPUT.fa</code> is read and a tree of depth <code>lmer_len</code> is constructed on the flight.</li>
<li><code>--idx &lt;INDEX_FILE&gt;:&lt;score&gt;</code>: in <code>INDEX_FILE</code> the tree structure was stored with <code>./makeTree</code>. A read will be considered to be a contamination if the score is bigger than <code>score</code>. The score is computed as the proportion of Lmers of a given read found in the tree. This method is very fast, every search is <code>O( 2 * Lmer * (L - Lmer + 1))</code> (the reverse complement has to be checked also), but has the drawback that is very memory intensive. This is why we have limited the construction of a tree to sequences whose size does not exceed 10 MB. Constructing a tree is very fast, that is why most of the times it might be not worth it to create the index file and then read it for every sample. Nevertheless, both options are provided.</li>
</ul>
</li>
<li><b>BLOOM</b>: this method is designed for large sequences up to 4GB. An index is precomputed, using <code>makeBloom</code>, so that it is computed only once for all samples. The index file name and the threshold score have to be specified through the following option:<ul>
<li><code>--idx &lt;INDEX_FILE&gt;:&lt;score&gt;</code> The score is computed as for the previous options.</li>
</ul>
</li>
</ul>
<h4>LowQ</h4>
<ul>
<li><code>--trimQ NO</code> or flag absent: nothing is done to the reads with low quality.</li>
<li><code>--trimQ ALL</code>: all reads containing at least one low quality nucleotide are redirected to <code>*_lowq.fq.gz</code></li>
<li><code>--trimQ ENDS</code>: look for low quality (below MINQ) base callings at the beginning and at the end of the read. Trim them at both ends until the quality is above the threshold. Keep the read in <code>*_good.fq.gz</code> and annotate in the fourth line where the read has been trimmed (starting to count from 0) if the length of the remaining part at least <code>MINL</code>. Redirect the read to <code>*_lowq.fq.gz</code> otherwise. Examples (-q 27 [&lt;] -m 25): <div class="fragment"><div class="line">ORIGINAL                                            FILTERED</div><div class="line">@ read 1081133                                      @ read 1081133</div><div class="line">GTACATTGATGAGCAACATGGGGCTTGAACTGGCGCTGAAACAGTTAGGA  GTACATTGATGAGCAACATGGGGCTTGAACTGGCGCTGAAACAGTTAGG</div><div class="line">+position: -144740                                  +position: -144740 TRIMQ:0:48</div><div class="line">IIIIIIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII9  IIIIIIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 3435                                         @ read  3435</div><div class="line">CAGTTCTTTGGTGTAGATGGAGCAGGACGGGATACCCATACCGTGACCCA  CTTTGGTGTAGATGGAGCAGGACGGGATACCCATACCGTG</div><div class="line">+position: -19377                                   +position: -19377  TRIMQ:5:44</div><div class="line">99999IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII99999  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 108110                                       @ read 108110</div><div class="line">CAGATAAATGCGAATGAATGGGTAACACTGGAGCTTTTAATGGCTGTAAC  AGATAAATGCGAATGAATGGGTAACACTGGAGCTTTTAATGGCTGTAA</div><div class="line">+position: -4142524                                 +position: -4142524  TRIMQ:1:48</div><div class="line">9IIIIIIIIIIIIIIIIIII9IIII9III9IIIIIIIIIIIIIIIIIII9  IIIIIIIIIIIIIIIIIII9IIII9III9IIIIIIIIIIIIIIIIIII</div><div class="line">@ read 108111</div><div class="line">AGCGTGACGGTGTAACGCCCGCCTTTTGATGACTGGGTTTCAAAGAAACG</div><div class="line">+position: -3336785</div><div class="line">999999999999999IIIIIIII9IIIIIIIII9II99999999999999</div></div><!-- fragment --></li>
<li><code>--trimQ FRAC [--percent p]</code>: redirect the read to <code>*_lowq.fq.gz</code> if there are more than <code>p%</code> nucleotides whose quality lies below the threshold. <code>p=5</code> per default. Examples (-q 27 [&lt;] -m 25 -p 5): <div class="fragment"><div class="line">ORIGINAL                                            FILTERED</div><div class="line">@ read 1081133                                      @ read 1081133</div><div class="line">GTACATTGATGAGCAACATGGGGCTTGAACTGGCGCTGAAACAGTTAGGA  GTACATTGATGAGCAACATGGGGCTTGAACTGGCGCTGAAACAGTTAGGA</div><div class="line">+position: -144740                                  +position: -144740</div><div class="line">IIIIIIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII9  IIIIIIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII9</div><div class="line">@ read 3435</div><div class="line">CAGTTCTTTGGTGTAGATGGAGCAGGACGGGATACCCATACCGTGACCCA</div><div class="line">+position: -19377</div><div class="line">99999IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII99999</div><div class="line">@ read 108110</div><div class="line">CAGATAAATGCGAATGAATGGGTAACACTGGAGCTTTTAATGGCTGTAAC</div><div class="line">+position: -4142524</div><div class="line">9IIIIIIIIIIIIIIIIIII9IIII9III9IIIIIIIIIIIIIIIIIII9</div><div class="line">@ read 108111</div><div class="line">AGCGTGACGGTGTAACGCCCGCCTTTTGATGACTGGGTTTCAAAGAAACG</div><div class="line">+position: -3336785</div><div class="line">999999999999999IIIIIIII9IIIIIIIII9II99999999999999</div></div><!-- fragment --></li>
<li><code>--trimQ ENDSFRAC --percent p</code>: first trim the ends as in the <code>ENDS</code> option. Accept the trimmed read if the number of low quality nucleotides does not exceed <code>p%</code> (default <code>p = 5</code>). Redirect the read to <code>*_lowq.fq.gz</code> otherwise. Examples (-q 27 [&lt;] -m 25 -p 5): Missing reads would land in <code>*_lowq.fq.gz</code>. <div class="fragment"><div class="line">ORIGINAL                                            FILTERED</div><div class="line">@ read 1081133                                      @ read 1081133</div><div class="line">GTACATTGATGAGCAACATGGGGCTTGAACTGGCGCTGAAACAGTTAGGA  GTACATTGATGAGCAACATGGGGCTTGAACTGGCGCTGAAACAGTTAGG</div><div class="line">+position: -144740                                  +position: -144740 TRIMQ:0:48</div><div class="line">IIIIIIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII9  IIIIIIIIIIIII9IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 3435                                         @ read  3435</div><div class="line">CAGTTCTTTGGTGTAGATGGAGCAGGACGGGATACCCATACCGTGACCCA  CTTTGGTGTAGATGGAGCAGGACGGGATACCCATACCGTG</div><div class="line">+position: -19377                                   +position: -19377  TRIMQ:5:44</div><div class="line">99999IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII99999  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 108110</div><div class="line">CAGATAAATGCGAATGAATGGGTAACACTGGAGCTTTTAATGGCTGTAAC</div><div class="line">+position: -4142524</div><div class="line">9IIIIIIIIIIIIIIIIIII9IIII9III9IIIIIIIIIIIIIIIIIII9</div><div class="line">@ read 108111</div><div class="line">AGCGTGACGGTGTAACGCCCGCCTTTTGATGACTGGGTTTCAAAGAAACG</div><div class="line">+position: -3336785</div><div class="line">999999999999999IIIIIIII9IIIIIIIII9II99999999999999</div></div><!-- fragment --></li>
<li><code>--trimQ GLOBAL --global n1:n2</code>: cut all reads globally <code>n1</code> nucleotides from the left and <code>n2</code> from the right.</li>
</ul>
<p><b>Note:</b> qualities are evaluated assuming the reads to follow the L - Illumina 1.8+ Phred+33, convention, see <a href="https://en.wikipedia.org/wiki/FASTQ_format#Encoding">Wikipedia</a>. Adjust the values for a different convention.</p>
<h4>N trimming</h4>
<p>We allow for the following options:</p>
<ul>
<li><code>--trimN NO</code> (or flag absent): Nothing is done to the reads containing N's.</li>
<li><code>--trimN ALL</code>: All reads containing at least one N are redirected to <code>*_NNNN.fq.gz</code></li>
<li><code>--trimN ENDS</code>: N's are trimmed if found at the ends, left "as is" otherwise. If the trimmed read length is smaller than MINL, it is discarded. Example: <div class="fragment"><div class="line">ORIGINAL                                           FILTERED</div><div class="line">@ read 1037                                        @ read 1037</div><div class="line">NNTCGACACAAGAAAATGCGCCAATTTTGAGCCAGACCCCAGTTACGCNN TCGACACAAGAAAATGCGCCAATTTTGAGCCAGACCCCAGTTACGC</div><div class="line">+position: -2044610                                +position: -2044610  TRIMN:2:47</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 1038</div><div class="line">AAAAAACGGTCGTGGTNGGCTACTGTTATTAAAGCGTTGGCTACAAAAAG</div><div class="line">+position: 1361068</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 1039</div><div class="line">CCAACGGACNATGGAAGTGCTNCGGCTGGNGTTTTTATCCTCCGGCATTC</div><div class="line">+position: -4282223</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 1043                                        @ read 1043</div><div class="line">AATCTGAAACGAAANCTGACAGCGNNCCCCGCTTCTGACAAAATAGGCGN AATCTGAAACGAAANCTGACAGCGNNCCCCGCTTCTGACAAAATAGGCG</div><div class="line">+position: -4685876                                +position: -4685876  TRIMN:0:48</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div></div><!-- fragment --></li>
<li><code>--trimN STRIP</code>: Obtain the largest N free subsequence of the read. Accept it if is at least MINL nucleotides long, redirect it to <code>*_NNNN.fq.gz</code> otherwise. Example: <div class="fragment"><div class="line">ORIGINAL                                           FILTERED</div><div class="line">@ read 1037                                        @ read 1037</div><div class="line">NNTCGACACAAGAAAATGCGCCAATTTTGAGCCAGACCCCAGTTACGCNN NNTCGACACAAGAAAATGCGCCAATTTTGAGCCAGACCCCAGTTACGCNN</div><div class="line">+position: -2044610                                +position: -2044610  TRIMN:2:47</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 1038                                        @ read 1038</div><div class="line">AAAAAACGGTCGTGGTNGGCTACTGTTATTAAAGCGTTGGCTACAAAAAG GGCTACTGTTATTAAAGCGTTGGCTACAAAAAG</div><div class="line">+position: 1361068                                 +position: 1361068  TRIMN:17:49</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 1039</div><div class="line">CCAACGGACNATGGAAGTGCTNCGGCTGGNGTTTTTATCCTCCGGCATTC</div><div class="line">+position: -4282223</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div><div class="line">@ read 1043</div><div class="line">AATCTGAAACGAAANCTGACAGCGNNCCCCGCTTCTGACAAAATAGGCGN</div><div class="line">+position: -4685876</div><div class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</div></div><!-- fragment --></li>
</ul>
<h2>Test/examples</h2>
<p>The examples in folder <code>examples/trimFilter_SReport/</code> work in the following way:</p>
<ol type="1">
<li>See folder <code>examples/fa_fq_files</code>. The file <code>EColi_rRNA.fq</code> was created with <code>create_fq.sh</code> and contains: <br />
 * 2e4 reads of length 50 from <code>EColi_genome.fa</code> with NO errors.<ul>
<li>5e3 reads of length 50 from <code>rRNA_modified.fa</code> with NO errrors (rRNA contaminations). <br />
 * Artificially generated reads with low quality score (see <code>create_fq.sh</code>)</li>
<li>Artificially generated reads with Ns (see <code>create_fq.sh</code>).</li>
<li>Adapter files: <code>adapter_even_long.fa</code>, <code>adapter_odd_long.fa</code>, <code>adapter_even_short.fa</code>, <code>adapter_odd_short.fa</code>. Fasta files containing one adapter sequence each, longer/shorter than 16 nucleotides and with an even/odd length.</li>
<li>Example files to test the adapter contamination searchs: <code>human_[even/odd]_wad_[even/odd]_[long/short].fq</code>. Short fastq files where adapters contaminations have been inserted in all possible ways: even/odd positions, at the beginning/middle/end of the reads. Read lengths are even or odd as the first suffix indicates. The adapter contaminations included are suggested by the second even/odd suffix, and the long/short suffix.</li>
</ul>
</li>
<li><code>adapters/run_example.sh</code>: runs examples of reads containing adapters contaminations. A set of different possibilities is covered. See README file inside the folder <code>adapters</code></li>
<li><code>run_example_TREE.sh</code>: the code was tested with flags: <div class="fragment"><div class="line">$ ../../bin/trimFilter -l 50 --ifq PATH/TO/EColi_rRNA.fq.gz </div><div class="line">--method TREE --ifa PATH/TO/rRNA_modified.fa:0.9:50 </div><div class="line">--trimQ ENDSFRAC --trimN STRIP -o tree</div></div><!-- fragment --> i.e., we check for contaminations from rRNA, trim reads with lowQ at the ends and less than 5% in the remaining part, and strip reads containing N's. The output should coincide with the files <code>example_TREE*</code> <br />
4. <code>run_example_BLOOM.sh</code>: <br />
 * bloom filter is generated for <code>rRNA_modified.fa</code> with FPR = 0.0075 and <code>kmersize=25</code>. The output should coincide with <code>rRNA_example.bf*</code>.<ul>
<li>trimFilter is run like in 2. but passing a bloom filter to look for contaminations with <code>score=0.4</code>.</li>
</ul>
</li>
<li>With this set up, it is possible to run further customized tests. <br />
 <b>NOTE:</b> <code>rRNA_modified.fa</code> is the <code>rRNA_CRUnit.fa</code> sequence, where we have <br />
 removed the lines containing N's for testing purposes. <br />
</li>
</ol>
<h2>Contributors</h2>
<p>Paula Pérez Rubio</p>
<h2>License</h2>
<p>GPL v3 (see LICENSE.txt) </p>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.14
</small></address>
</body>
</html>
